### PR TITLE
feat(switch): support variant mode

### LINKED ISSUE #1550: (Switch): Add new appearance
# Feature request 🚀
To be supplemented later

### PR DESCRIPTION
**Pull request by YHoney7**
## Checklist

List of tasks you have already done and plan to do.

- [ ] Fix linting errors
- [ ] Tests have been added / updated (or snapshots)

## Change information

Describe your modifications here.

## Issues

close #1550 

## Related Links

Links related to this pr.

### CODE REVIEW

**Comment by haoziqaq** [comment]
[playground](https://varlet-ui-playground-8f8zpcqk7-varletjs.vercel.app/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgbXNnID0gcmVmKCdIZWxsbyBWYXJsZXQhJylcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDx2YXItc3dpdGNoIHR5cGU9XCJwcmltYXJ5XCIgdmFyaWFudCAvPlxuPC90ZW1wbGF0ZT5cbiIsIkFwcFdyYXBwZXIudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBBcHAgZnJvbSAnLi9BcHAudnVlJ1xuaW1wb3J0IHsgaW5zdGFsbFZhcmxldFVJIH0gZnJvbSAnLi92YXJsZXQtcmVwbC1wbHVnaW4uanMnXG5cbmluc3RhbGxWYXJsZXRVSSgpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8QXBwIC8+XG48L3RlbXBsYXRlPlxuIiwidmFybGV0LXJlcGwtcGx1Z2luLmpzIjoiaW1wb3J0IFZhcmxldFVJIGZyb20gJ0B2YXJsZXQvdWknXG5pbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5cbmNvbnN0IHZhcmxldENzcyA9ICcuL3ZhcmxldC5jc3MnXG5jb25zdCB2YXJsZXRUb3VjaEVtdWxhdG9yID0gJy4vdmFybGV0LXRvdWNoLWVtdWxhdG9yLmpzJ1xuXG5hd2FpdCBhcHBlbmRTdHlsZSgpXG5cbmV4cG9ydCBmdW5jdGlvbiBpbnN0YWxsVmFybGV0VUkoKSB7XG4gIGNvbnN0IHsgcGFyZW50IH0gPSB3aW5kb3dcblxuICBjb25zdCBzdHlsZSA9IGRvY3VtZW50LmNyZWF0ZUVsZW1lbnQoJ3N0eWxlJylcbiAgc3R5bGUuaW5uZXJIVE1MID0gYFxuICAgIGJvZHkge1xuICAgICAgbWluLWhlaWdodDogMTAwdmg7XG4gICAgICBwYWRkaW5nOiAxNnB4O1xuICAgICAgbWFyZ2luOiAwO1xuICAgICAgY29sb3I6IHZhcigtLWNvbG9yLXRleHQpO1xuICAgICAgYmFja2dyb3VuZC1jb2xvcjogdmFyKC0tY29sb3ItYm9keSk7XG4gICAgfVxuXG4gICAgKjo6LXdlYmtpdC1zY3JvbGxiYXIge1xuICAgICAgZGlzcGxheTogbm9uZTtcbiAgICB9XG4gIGBcbiAgZG9jdW1lbnQuaGVhZC5hcHBlbmRDaGlsZChzdHlsZSlcblxuICBjb25zdCBzY3JpcHQgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdzY3JpcHQnKVxuICBzY3JpcHQuc3JjID0gdmFybGV0VG91Y2hFbXVsYXRvclxuICBkb2N1bWVudC5ib2R5LmFwcGVuZENoaWxkKHNjcmlwdClcblxuICBjb25zdCB0aGVtZU1hcCA9IHtcbiAgICBsaWdodFRoZW1lOiBudWxsLFxuICAgIGRhcmtUaGVtZTogVmFybGV0VUkuVGhlbWVzLmRhcmssXG4gICAgbWQzTGlnaHRUaGVtZTogVmFybGV0VUkuVGhlbWVzLm1kM0xpZ2h0LFxuICAgIG1kM0RhcmtUaGVtZTogVmFybGV0VUkuVGhlbWVzLm1kM0RhcmssXG4gIH1cblxuICBjb25zdCB0aGVtZSA9IHBhcmVudC5kb2N1bWVudC5kb2N1bWVudEVsZW1lbnQuZ2V0QXR0cmlidXRlKCd0aGVtZScpXG5cbiAgaWYgKHRoZW1lKSB7XG4gICAgVmFybGV0VUkuU3R5bGVQcm92aWRlcih0aGVtZU1hcFt0aGVtZV0pXG4gIH1cblxuICB3aW5kb3cuYWRkRXZlbnRMaXN0ZW5lcignbWVzc2FnZScsICh7IGRhdGEgfSkgPT4ge1xuICAgIGlmIChkYXRhLmFjdGlvbiA9PT0gJ3RoZW1lLWNoYW5nZScpIHtcbiAgICAgIFZhcmxldFVJLlN0eWxlUHJvdmlkZXIodGhlbWVNYXBbZGF0YS52YWx1ZV0pXG4gICAgfVxuICB9KVxuXG4gIGRvY3VtZW50LmFkZEV2ZW50TGlzdGVuZXIoJ2NsaWNrJywgKCkgPT4ge1xuICAgIHdpbmRvdy5wYXJlbnQuZG9jdW1lbnQuZGlzcGF0Y2hFdmVudChuZXcgRXZlbnQoJ2NsaWNrJykpXG4gIH0pXG5cbiAgY29uc3QgaW5zdGFuY2UgPSBnZXRDdXJyZW50SW5zdGFuY2UoKVxuICBpbnN0YW5jZS5hcHBDb250ZXh0LmFwcC51c2UoVmFybGV0VUkpXG59XG5cbmV4cG9ydCBmdW5jdGlvbiBhcHBlbmRTdHlsZSgpIHtcbiAgcmV0dXJuIG5ldyBQcm9taXNlKChyZXNvbHZlLCByZWplY3QpID0+IHtcbiAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgbGluay5yZWwgPSAnc3R5bGVzaGVldCdcbiAgICBsaW5rLmhyZWYgPSB2YXJsZXRDc3NcbiAgICBsaW5rLm9ubG9hZCA9IHJlc29sdmVcbiAgICBsaW5rLm9uZXJyb3IgPSByZWplY3RcbiAgICBkb2N1bWVudC5oZWFkLmFwcGVuZENoaWxkKGxpbmspXG4gIH0pXG59IiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwiQHZhcmxldC91aVwiOiBcIi4vdmFybGV0LmVzbS5qc1wiLFxuICAgIFwiQHZhcmxldC91aS9qc29uL2FyZWEuanNvblwiOiBcIi4vdmFybGV0LWFyZWEuanNcIlxuICB9XG59IiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlLFxuICAgIFwianN4XCI6IFwiUHJlc2VydmVcIixcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn0ifQ==) 显示似乎有些问题。

**Comment by haoziqaq** [comment]
还有一个问题就是 ripple 似乎有点过于大了。

**Comment by haoziqaq** [comment]
props.size 似乎在 variant 模式下没有适配

**Comment by YHoney7** [comment]
我看下playground是有问题, ripple的尺寸目前是对比figma标注的尺寸的

File: packages/varlet-ui/src/switch/Switch.vue [hunk 1]
```diff
@@ -96,8 +96,9 @@ export default defineComponent({
     const { bindForm, form } = useForm()
     const { errorMessage, validateWithTrigger: vt, validate: v, resetValidation } = useValidation()
     const { hovering, handleHovering } = useHoverOverlay()
+    // const variant = computed<boolean | undefined>(() => props.variant)
```
Comment (BeADre) [inline packages/varlet-ui/src/switch/Switch.vue:99]: 移除无用的注释

File: packages/varlet-ui/src/switch/docs/en-US.md [hunk 2]
```diff
@@ -143,6 +157,7 @@ function handleBeforeChange(value, change) {
 | `rules`          | Validation rules | _Array<(value: any) => any>_  | `-` |
 | `lazy-change`    | Whether to allow the `before-change` event to be triggered | _boolean_  | `false` |
 | `validate-trigger` | Timing to trigger verification, optional values are `onChange` `onLazyChange` | _ValidateTriggers[]_ | `['onChange', 'onLazyChange']` |
+| `variant` ***3.2.0*** | Variant mode | _boolean_ | `false` |
```
Comment (BeADre) [inline packages/varlet-ui/src/switch/docs/en-US.md:160]: @haoziqaq 现在新增的属性后面有必要跟上版本号吗？

File: packages/varlet-ui/src/switch/Switch.vue (hunk 1 above)
Comment (YHoney7) [reply packages/varlet-ui/src/switch/Switch.vue:99]: 好的，现在还是有些问题，还在更改

File: packages/varlet-ui/src/switch/docs/en-US.md (hunk 2 above)
Comment (haoziqaq) [reply packages/varlet-ui/src/switch/docs/en-US.md:160]: 要跟上，不过理论上应该是 3.2.1

**Comment by BeADre** [comment]
参考[md3](https://m3.material.io/components/switch/overview), 间距还有样式好像不太对

**Comment by YHoney7** [comment]
image
我们看到的效果似乎不一样

**Comment by YHoney7** [comment]
image
@BeADre 分为两种形式，目前实现的是没有图标的，但当前这个看来还缺个hover和拖动实现开关的效果

**Comment by YHoney7** [comment]
收到👌

**Comment by chouchouji** [comment]
影响到了form的快照，看下为啥，如果有必要就更新下快照

**Comment by YHoney7** [comment]
好的

**Comment by BeADre** [comment]
样式上还有点小问题，[官方](https://m3.material.io/components/switch/accessibility#c0e9fae1-48df-428b-b028-4f7be071ada3)是hover的时候只有水波纹，按下去里面的滑块才会变大，还有一个就是亮色模式并打开switch的时候hover效果会导致滑块颜色改变，这里颜色应该是不会变化的

**Comment by YHoney7** [comment]
收到, 已经在改进了

File: packages/varlet-ui/src/switch/example/index.vue [hunk 3]
```diff
@@ -51,4 +52,6 @@ onThemeChange()
   <var-space :size="['3vmin', '9vmin']">
     <var-switch v-model="value5" lazy-change @before-change="handleBeforeChange" />
   </var-space>
+  <app-type>{{ t('variant') }}</app-type>
+  <var-switch variant disabled v-model="value6" />
```
Comment (chouchouji) [inline packages/varlet-ui/src/switch/example/index.vue:56]: 示例开启了disabled，文档没有

**Comment by BeADre** [comment]
我点击switch其他的地方，switch的块应该也会变大，这是官方的[实现](https://material-web.dev/components/switch/stories/)

**Comment by YHoney7** [comment]
明白， 我看了你发的示例，在开和关的情况下还有个小回弹的效果，但是在另个[示例](https://m3.material.io/components/switch/accessibility)中却没有这个回弹效果

**Review by BeADre — APPROVED** [review]

**Review by BeADre — APPROVED** [review]

**Pull request merged**